### ISSUE TITLE
storageSpec.VolumeClaim is not respected

### ISSUE THREAD

**Issue by auhlig** [coreos/prometheus-operator#2535]
**What happened?**
Configured a Prometheus with
```
storage:
  volumeClaimTemplate:
    metadata:
      name: "k8s-prometheus"
    spec:
      accessModes:
        - "ReadWriteMany"
      resources:
        requests:
          storage: "100Gi"
```

However, the operator ignores the VolumeClaimTemplate and creates the PVC with `spec.accessModes: ["ReadWriteOnce"]`.

Cause: The access mode is hardcoded [here](https://github.com/coreos/prometheus-operator/blob/master/pkg/prometheus/statefulset.go#L213).

**Did you expect to see some different?**

Operator creates PVC according to pvc template.

**Environment**

* Prometheus Operator version:

    `v0.29.0`

**Comment by squat** [comment]
@auhlig we intentionally hard code the accessModes to `ReadWriteOnce` because no two Prometheus servers should ever be sharing their storage; this will only cause problems. What is your use-case or need for `ReadWriteMany`?

**Comment by auhlig** [comment]
Thanks for the answer @squat. 
I agree that if multiple Promtheis share storage causes mayhem.

However, `StorageSpec.VolumeClaimTemplate` is a `v1.PersistentVolumeClaim` and it should be up to the user to configure that. 
Currently, the configured PVC template is just silently ignored. This is quite confusing. 

I propose to leave configuration up to the user and use the volume claim template as configured in the storage spec. Potentially adding a note to the documentation that a persistent volume shall only be used by one Prometheus to avoid data corruption.

**Comment by squat** [comment]
@auhlig I don't think that is a good idea. I suspect that the result of making that change will be user bug reports about corrupted data. There is not a valid reason for using ReadWriteMany in Prometheus Volumes and the operator should enforce this view.

Consider what would happen if a user ran a multi-replica Prometheus Statefulset, as is the default in kube-prometheus, but with a ReadWriteMany PVC:
1. two identical PVCs would be creating requesting storage
2. one PV would be provisioned automatically as it could satisfy both PVCs
3. the single PV would be mounted by both Prometheus servers
4. the servers would corrupt each-other's data

**Comment by auhlig** [comment]
Thanks for the answer @squat.
I understand the implications of setting `ReadWriteMany` on the PVC and I completely agree that one should be using `ReadWriteOnce` to avoid corrupted data.

Unfortunately, we have to use pv's with `accessMode=ReadWriteMany` in some of our clusters. 
We're aware that this is not ideal and we're working on fixing that. 
However, there are no other option for us than to use the existing volumes with the `RWX` mode.
This stops us from using the prometheus operator right now :(

Given your feedback I propose:
Default to `accessMode=ReadWriteOnce`. 
Only if the user explicitly sets a different accessMode via the volume claim template it should be overwritten. 
Add an explicit warning for choosing a different access mode to the documentation describing the implications.

**Comment by brancz** [comment]
I think I agree with @squat, I think the Prometheus Operator should give hard guidance on the best way to run Prometheus and if the cluster is unable to deliver that, then the cluster "incompatible" with the best practices.

**Comment by BugRoger** [comment]
Considering how much functionality this operator covers, I think, less opinion and more flexibility would be beneficial. This flag basically rules out the usage of this operator on our whole fleet of clusters. We either have to fork or maintain or own solution. It's a lose/lose situation. We have to spend our efforts elsewhere and the operator doesn't get contributions.

**Comment by jobrs** [comment]
I would agree with @squat if only Kubernetes would fall-back to RWX volumes to satisfy RWO claims (if RWO volumes are not available). But as long as this is not the case, I find hardcoding not an adequate solution especially since you are reusing VolumeClaimTemplate.

**Comment by metalmatze** [comment]
I agree with @squat and @brancz on their previous comments.
Enforcing the PVC mode to `ReadWriteOnce` has multiple advantages:
1. Users are safeguarded to not get themselves into trouble as easily
2. Thus we, as maintainers, do **not** get as many bug reports from users who unknowingly use `ReadWriteMany`.
3. Try to be opinionated, not with our own opinion, but with what's recommended by Prometheus itself.

Just like NFS, I'd argue that running `ReadWriteMany` is a bad idea when it comes to storage and Prometheus.

Patching the Prometheus Operator to support `ReadWriteMany` is just a few lines of code and once that's changed you can build a new image with `make operator image`.

As you said it would be a workaround and doesn't solve the real problem.
> Unfortunately, we have to use pv's with accessMode=ReadWriteMany in some of our clusters.
We're aware that this is not ideal and we're working on fixing that.

**Comment by auhlig** [comment]
Thanks a lot for taking the time @metalmatze, @brancz, @squat.

  > 1. Users are safeguarded to not get themselves into trouble as easily
  > 2. Thus we, as maintainers, do not get as many bug reports from users who unknowingly use ReadWriteMany.

I completely agree with you. 
Users as well as maintainers shouldn't be bothered by an unknowingly set configuration.
However, the reason I had to open this issue is exactly that.
Storage can currently be configured via a volume claim template. 
The operator overrides some parameters of my template without me knowing nor wanting this. 
IMO this is just confusing.

Moreover, I also agree that it's a good idea to provide guidelines and useful defaults on how Prometheus is deployed. 
Nevertheless, this shouldn't prevent users from applying the configuration they want.
In this case this would mean setting `ReadWriteOnce` as a default.
Users can override this knowingly and at their own risk.
The documentation should be updated stating the implications.

**Comment by ant31** [comment]
>Nevertheless, this shouldn't prevent users from applying the configuration they want.
The goal of the project is to simplify as much as possible running prometheus on Kubernetes using the best practices. Including a way to corrupt user data is against this goal. 

As alternative, we could NOT use the upstream `v1.volumeClaimTemplate` and create new struct without the accessMode field, that way it would raise a validatationError if someone misconfigure it. 
It could remove probably some of the confusion.

**Comment by brancz** [comment]
Closing due to inactivity.

**Comment by midnightconman** [comment]
I am not sure if commenting on a closed issue is better than opening a new one... so I'll start here. If you guys want me to open a new issue, let me know.

I think we are confusing a couple of things here... The access mode of the PVC doesn't always mean that it will be accessed by multiple actors, especially when the volumeClaim is being used by a statefulset. When you have a volumeClaimTemplate on a statefulset, the statefulset will still create a separate volume per replica even if you use RWX. Along these lines, if I have a storageclass that requires that my PVs be of type RWX, I can't use that storageclass with the prometheus-operator. This doesn't make too much sense to me, as having a PVC with RWX doesn't mean multiple pods will try to use the same underlying storage space. Does that make sense?

TLDR;
It's not possible for someone to use a PVC with access mode RWX for two prometheus pod with the prometheus operator, since a different PVC is created for each replica for each prometheus cluster.
